exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/BaseRootExec.java

```diff
@@ -184,4 +179,38 @@ public void close() throws Exception {
       }
     }
   }
+
+  /**
+   * Given a list of operators and a stack trace, walks the stack trace and
+   * the operator list to find the leaf-most operator, which is the one
+   * that was active when the exception was thrown. Handle the cases in
+   * which no operator was active, each operator had multiple methods on
+   * the stack, or the exception was thrown in some class called by
+   * the operator.
+   * <p>
+   * Not all operators leave a mark in the trace. In particular if a the
+   * call stack is only through base-class methods, then we have no way to
+   * know the actual class during the call. This is OK because the leaf
+   * methods are just pass-through operations, they are unlikely to fail.
+   *
+   * @param <T> the type of the operator. Parameterized to allow easier
+   * testing
+   * @param dag the list of operators from root-most to leaf-most
+   * @param e the exception thrown somewhere in the operator tree
+   * @return the leaf-most operator, if any
+   */
+  public static <T> T findLeaf(List<T> dag, Throwable e) {
+    StackTraceElement[] trace = e.getStackTrace();
```

Comment:
I see the order is important for the outer loop 
```java 
for (int i = dag.size() - 1; i >= 0; i--) {
...
}
```
 but in an inner loop, you're iterating for each stack trace element so the suggested ```Set<String>``` won't break your algorithm. Though I don't force the change since we know that operators and stack trace collections are small.